doc/mpls/MPLS_hld.md

```diff
@@ -0,0 +1,563 @@
+###### Functions
+The following are new functions for RouteOrch:
+```
+  /* Consumer handler for all events in APPL_DB LABEL_ROUTE_TABLE */
+  void RouteOrch::doLabelTask(Consumer& consumer);
+  /* Handler to process new MPLS route from LABEL_ROUTE_TABLE */
+  bool addLabelRoute(LabelRouteBulkContext& ctx, const NextHopGroupKey&);
+  /* Handler to process MPLS route removal from LABEL_ROUTE_TABLE */
+  bool removeLabelRoute(LabelRouteBulkContext& ctx);
+```
+
+##### NeighOrch
+NeighOrch is an existing component of the OrchAgent daemon in the SWSS container.  NeighOrch monitors operations on Neighbor related tables in APPL_DB.  Additionally NeighOrch coordinates Next Hop operations with RouteOrch and converts operations into SAI commands to manage Next Hop objects.
+
+For MPLS, NeighOrch has been extended to send create/remove SAI requests for MPLS NextHop objects (ie, NextHop objects that include non-empty LabelStack information) when associated Neighbor objects are created/removed. This MPLS NextHop behavior parallels the existing IPv4/IPv6 NextHop behavior in NeighOrch.
+###### Functions
+Existing functions from NeighOrch are updated to include NextHopKey parameter instead of IpAddress and visibility is raised to public for RouteOrch accessibility.
+```
+    bool hasNextHop(const NextHopKey&);
+    bool addNextHop(const NextHopKey&);
+    bool removeNextHop(const NextHopKey&);
+```
+
+##### CrmOrch
+CrmOrch is an existing component of the OrchAgent daemon in the SWSS container.  CrmOrch monitors resource usage in the SONiC system and triggers alarms when configurable thresholds are reached.
+
+For MPLS, CrmOrch has been extended to monitor the number of MPLS in-segment entries against the platform-specific number of entries available.  To facilitate this, a new CRM resource type CRM_MPLS_INSEG has been added to CrmOrch and mapped to the existing SAI object type SAI_OBJECT_TYPE_INSEG_ENTRY for querying via sai_object_type_get_availability().  This MPLS behavior parallels the existing IPv4/IPv6 behavior with CRM_IPV4_ROUTE/SAI_SWITCH_ATTR_AVAILABLE_IPV4_ROUTE_ENTRY and CRM_IPV6_ROUTE/SAI_SWITCH_ATTR_AVAILABLE_IPV6_ROUTE_ENTRY.
+###### Functions
+No new functions were required for CrmOrch MPLS support.
+
+##### Label/LabelStack
+Label and LabelStack are new type utilities of the OrchAgent daemon in the SWSS container.
+These types are introduced to represent the MPLS label or label stack when found in an in-segment entry or next hop.
+```
+typedef uint32_t Label;
+class LabelStack
+{
+    ... class definition abbreviated ...
+private:
+    std::set<Label> m_labelstack;
+};
+```
+##### NextHopKey
+NextHopKey is an existing utility of the OrchAgent daemon in the SWSS container.  NextHopKey is used by both RouteOrch and NeighOrch to coordinate Next Hop operations.
+
+For MPLS, the NextHopKey struct is modified to include a LabelStack field.
+```
+struct NextHopKey
```

Comment:
Suggest have a constructor to specify this as mpls nexthop similar to current `overlay_nh`